ooni/nettests/blocking/web_connectivity.py

```diff
@@ -0,0 +1,512 @@
+from twisted.python import usage
+
+from ooni import geoip
+from ooni.utils import log
+
+from ooni.backend_client import WebConnectivityClient
+
+from ooni.utils.net import StringProducer, BodyReceiver
+from ooni.templates import httpt, dnst
+from ooni.errors import failureToString
+
+REQUEST_HEADERS = {
+    'User-Agent': ['Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, '
+                   'like Gecko) Chrome/47.0.2526.106 Safari/537.36'],
+    'Accept-Language': ['en-US;q=0.8,en;q=0.5'],
+    'Accept': ['text/html,application/xhtml+xml,application/xml;q=0.9,'
+               '*/*;q=0.8']
+}
+
+class InvalidControlResponse(Exception):
+    pass
+
+class TCPConnectProtocol(Protocol):
+    def connectionMade(self):
+        self.transport.loseConnection()
+
+class TCPConnectFactory(Factory):
+    noisy = False
+    def buildProtocol(self, addr):
+        return TCPConnectProtocol()
+
+
+class UsageOptions(usage.Options):
+    optParameters = [
+        ['url', 'u', None, 'Specify a single URL to test'],
+        ['dns-discovery', 'd', 'whoami.akamai.net', 'Specify the dns discovery test helper'],
+        ['backend', 'b', None, 'The web_consistency backend test helper'],
+    ]
+
+
+def is_public_ipv4_address(address):
+    try:
+        ip_address = IPv4Address(address)
+        if not any([ip_address.is_private,
+                    ip_address.is_loopback]):
+            return True
+        return False
+    except AddressValueError:
+        return None
+
+class WebConnectivityTest(httpt.HTTPTest, dnst.DNSTest):
+    """
+    Web connectivity
+    """
+    name = "Web connectivity"
+    description = ("Performs a HTTP GET request over Tor and one over the "
+                  "local network and compares the two results.")
+    author = "Arturo Filastò"
+    version = "0.1.0"
+
+    contentDecoders = [('gzip', GzipDecoder)]
+
+    usageOptions = UsageOptions
+
+    inputFile = [
+        'file', 'f', None, 'List of URLS to perform GET requests to'
+    ]
+
+    requiredTestHelpers = {
+        'backend': 'web-connectivity',
+        'dns-discovery': 'dns-discovery'
+    }
+    requiredOptions = ['backend', 'dns-discovery']
+    requiresRoot = False
+    requiresTor = False
+    followRedirects = True
+
+    # Factor used to determine HTTP blockpage detection
+    # the factor 0.7 comes from http://www3.cs.stonybrook.edu/~phillipa/papers/JLFG14.pdf
+    factor = 0.7
+    resolverIp = None
+
+    @classmethod
+    @defer.inlineCallbacks
+    def setUpClass(cls):
+        try:
+            answers = yield client.lookupAddress(
+                cls.localOptions['dns-discovery']
+            )
+            assert len(answers) > 0
+            assert len(answers[0]) > 0
+            cls.resolverIp = answers[0][0].payload.dottedQuad()
+        except Exception as exc:
+            log.exception(exc)
+            log.err("Failed to lookup the resolver IP address")
+
+
+    def inputProcessor(self, filename):
```

Comment:
if we're reading in multiple urls, 2 things that would be nice to have parameterized / be able to do:
1. randomize the order urls are tested
2. add a delay between tests to let the network recover